File: clients/python/cutout.py

"""Project Cutout — minimal Python client (stdlib only).

Usage:
    from cutout import Client

    bus = Client(agent_id="koda")  # reads CUTOUT_URL / CUTOUT_TOKEN
    msg = bus.post_message(thread_id="demo-thread", from_="koda",
                           to="instinct", type="question",
                           body="Retest passed 7/7. Ready to ship?",
                           idempotency_key="qa-handoff-001")
    batch = bus.get_messages(wait=30)          # long-poll for replies
    bus.post_receipt(msg["id"], agent="koda", status="acted")
    bus.resolve_thread(thread_id="demo-thread", from_="koda")
"""

import json
import os
import urllib.parse
import urllib.request


class CutoutError(Exception):
    def __init__(self, status, body):
        self.status = status
        self.body = body
        super().__init__("cutout: HTTP %s: %s" % (status, body))


class Client:
    def __init__(self, base_url=None, token=None, agent_id=None,
                 timeout=75):
        self.base_url = (base_url or os.environ.get(
            "CUTOUT_URL", "http://127.0.0.1:8765")).rstrip("/")
        self.token = token or os.environ.get("CUTOUT_TOKEN")
        if not self.token:
            raise CutoutError(0, "set CUTOUT_TOKEN")
        self.agent_id = agent_id or os.environ.get("CUTOUT_AGENT_ID")
        self.timeout = timeout

    def _request(self, method, path, body=None, params=None, timeout=None):
        url = self.base_url + path
        if params:
            url += "?" + urllib.parse.urlencode(
                {k: v for k, v in params.items() if v is not None})
        data = json.dumps(body).encode("utf-8") if body is not None else None
        req = urllib.request.Request(url, data=data, method=method)
        req.add_header("Authorization", "Bearer " + self.token)
        if self.agent_id:
            req.add_header("X-Agent-Id", self.agent_id)
        if data:
            req.add_header("Content-Type", "application/json")
        try:
            with urllib.request.urlopen(
                    req, timeout=timeout or self.timeout) as resp:
                raw = resp.read().decode("utf-8")
                return json.loads(raw) if raw else {}
        except urllib.error.HTTPError as exc:
            raise CutoutError(exc.code, exc.read().decode("utf-8",
                                                               "replace"))

    # -- API -----------------------------------------------------------

    def post_message(self, thread_id, from_, to, type, body,
                     reply_to=None, metadata=None, idempotency_key=None):
        payload = {"thread_id": thread_id, "from": from_, "to": to,
                   "type": type, "body": body}
        if reply_to:
            payload["reply_to"] = reply_to
        if metadata is not None:
            payload["metadata"] = metadata
        if idempotency_key:
            # safe to retry: the server returns the original message
            # with "duplicate": true instead of appending again
            payload["idempotency_key"] = idempotency_key
        return self._request("POST", "/v1/messages", body=payload)

    def resolve_thread(self, thread_id, from_, to="*", body="Resolved.",
                       metadata=None):
        """Mark a thread resolved (reopened by any later message)."""
        return self.post_message(thread_id, from_, to, "resolve", body,
                                 metadata=metadata)

    def get_messages(self, since=None, thread_id=None, to=None,
                     wait=0, limit=50):
        # long-poll: allow the server up to `wait` seconds plus headroom
        timeout = self.timeout if wait == 0 else wait + 15
        return self._request("GET", "/v1/messages",
                             params={"since": since, "thread_id": thread_id,
                                     "to": to, "wait": wait, "limit": limit},
                             timeout=timeout)

    def post_receipt(self, message_id, agent, status):
        return self._request("POST", "/v1/receipts",
                             body={"message_id": message_id,
                                   "agent": agent, "status": status})

    def get_threads(self):
        return self._request("GET", "/v1/threads")

File: clients/node/cutout.js

/**
 * Project Cutout — minimal Node.js client (no dependencies, Node 18+).
 *
 * Usage:
 *   import { CutoutClient } from "./cutout.js";
 *
 *   const bus = new CutoutClient({ agentId: "instinct" });
 *   // reads CUTOUT_URL / CUTOUT_TOKEN from the environment
 *   const msg = await bus.postMessage({
 *     threadId: "demo-thread", from: "instinct", to: "koda",
 *     type: "note", body: "Got it — reviewing now.",
 *     idempotencyKey: "review-ack-001",
 *   });
 *   const batch = await bus.getMessages({ wait: 30 }); // long-poll
 *   await bus.postReceipt({ messageId: msg.id, agent: "instinct",
 *                           status: "acted" });
 *   await bus.resolveThread({ threadId: "demo-thread",
 *                             from: "instinct" });
 */

export class CutoutError extends Error {
  constructor(status, body) {
    super(`cutout: HTTP ${status}: ${body}`);
    this.name = "CutoutError";
    this.status = status;
    this.body = body;
  }
}

export class CutoutClient {
  constructor({ baseUrl, token, agentId, timeoutMs } = {}) {
    this.baseUrl = (baseUrl || process.env.CUTOUT_URL ||
                    "http://127.0.0.1:8765").replace(/\/+$/, "");
    this.token = token || process.env.CUTOUT_TOKEN;
    if (!this.token) throw new CutoutError(0, "set CUTOUT_TOKEN");
    this.agentId = agentId || process.env.CUTOUT_AGENT_ID || null;
    this.timeoutMs = timeoutMs || 75000;
  }

  async _request(method, path, { body, params, timeoutMs } = {}) {
    let url = this.baseUrl + path;
    if (params) {
      const qs = new URLSearchParams();
      for (const [k, v] of Object.entries(params)) {
        if (v !== undefined && v !== null) qs.append(k, String(v));
      }
      const s = qs.toString();
      if (s) url += "?" + s;
    }
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(),
                             timeoutMs || this.timeoutMs);
    try {
      const resp = await fetch(url, {
        method,
        headers: {
          "Authorization": "Bearer " + this.token,
          ...(this.agentId ? { "X-Agent-Id": this.agentId } : {}),
          ...(body !== undefined
              ? { "Content-Type": "application/json" } : {}),
        },
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: ctrl.signal,
      });
      const text = await resp.text();
      if (!resp.ok) throw new CutoutError(resp.status, text);
      return text ? JSON.parse(text) : {};
    } catch (err) {
      if (err instanceof CutoutError) throw err;
      throw new CutoutError(0, String(err && err.message || err));
    } finally {
      clearTimeout(timer);
    }
  }

  // -- API ------------------------------------------------------------

  postMessage({ threadId, from, to, type, body, replyTo, metadata,
                idempotencyKey }) {
    const payload = { thread_id: threadId, from, to, type, body };
    if (replyTo) payload.reply_to = replyTo;
    if (metadata !== undefined) payload.metadata = metadata;
    // safe to retry: the server returns the original message with
    // "duplicate": true instead of appending again
    if (idempotencyKey) payload.idempotency_key = idempotencyKey;
    return this._request("POST", "/v1/messages", { body: payload });
  }

  resolveThread({ threadId, from, to = "*", body = "Resolved.",
                  metadata }) {
    // Mark a thread resolved (reopened by any later message).
    return this.postMessage({ threadId, from, to, type: "resolve",
                              body, metadata });
  }

  getMessages({ since, threadId, to, wait = 0, limit = 50 } = {}) {
    // long-poll: allow the server `wait` seconds plus headroom
    const timeoutMs = wait === 0 ? this.timeoutMs : (wait + 15) * 1000;
    return this._request("GET", "/v1/messages", {
      params: { since, thread_id: threadId, to, wait, limit },
      timeoutMs,
    });
  }

  postReceipt({ messageId, agent, status }) {
    return this._request("POST", "/v1/receipts", {
      body: { message_id: messageId, agent, status },
    });
  }

  getThreads() {
    return this._request("GET", "/v1/threads");
  }
}
